torch/package/module_environment.py

```diff
@@ -0,0 +1,219 @@
+        if importers is None:
+            importers = [DefaultImporter]
+        self._importers = importers
+
+    def import_module(self, module_name: str) -> ModuleType:
+        """Import 'module_name' from this environment.
+
+        Importers will be tried in order of `self.importers`, and the first match will be taken.
+        """
+        last_err = None
+        for importer in self._importers:
+            if not isinstance(importer, BaseImporter):
+                raise TypeError(
+                    f"{importer} is not a BaseImporter. "
+                    "All importers in ModuleEnv must inherit from BaseImporter."
+                )
+            try:
+                return importer.import_module(module_name)
+            except ModuleNotFoundError as err:
+                last_err = err
+
+        if last_err is not None:
+            raise last_err
+        else:
+            raise ModuleNotFoundError(module_name)
+
+    def get_name(
+        self, obj: Any, name: Optional[str] = None, check_obj: bool = True
+    ) -> Tuple[str, str]:
+        """Given an object, return a name that can be used to retrieve the
+        object from this environment.
+
+        Args:
+            obj: An object to get the the module-environment-relative name for.
+            name: If set, use this name instead of looking up __name__ or __qualname__ on `obj`.
+                This is only here to match how Pickler handles __reduce__ functions that return a string,
+                don't use otherwise.
+            check_obj: If true, check that retrieving the returned name produces the same object as `obj`.
+                This should only be false if you plan to call `check_obj()` somewhere else.
+
+        Returns:
+            A tuple (parent_module_name, attr_name) that can be used to retrieve `obj` from this environment.
+            Use it like:
+                mod = module_env.import_module(parent_module_name)
+                obj = getattr(mod, attr_name)
+
+        Raises:
+            ModuleEnvError: `check_obj()` has failed.
+        """
+        if name is None:
+            name = getattr(obj, "__qualname__", None)
+        if name is None:
+            name = obj.__name__
+
+        orig_module_name = self._whichmodule(obj, name)
+        # Demangle the module name before importing. If this obj came out of a
+        # PackageImporter, `__module__` will be mangled. See mangling.md for
+        # details.
+        module_name = demangle(orig_module_name)
+
+        if check_obj:
+            module = self.import_module(module_name)
+            obj2, _ = _getattribute(module, name)
+            self.check_obj(obj, obj2, name)
+
+        return module_name, name
+
+    def is_default(self) -> bool:
+        """Returns true if this ModuleEnv only represents the default Python environment."""
+        return self._importers == [DefaultImporter]
+
+    def check_obj(self, obj, obj2, name=None):
```

Comment:
This API is weird as well. It feels way to specialized to the pickler. I think all the functionality can be encapsulated in `get_name`.